llm_studio/src/losses/text_causal_language_modeling_losses.py

```diff
@@ -25,6 +24,33 @@ def forward(self, logits, labels):
         return self.loss_fn(shift_logits, shift_labels)
 
 
+class LengthBasedTACE(nn.Module):
+    def __init__(self, cfg: Any, length_penalty_coeff: float = 0.1):
+        super().__init__()
+        self.cfg = cfg
+        self.loss_fn = nn.CrossEntropyLoss()
+        self.length_penalty_coeff = length_penalty_coeff
+
+    def forward(self, logits, labels):
+        shift_logits = logits[..., :-1, :].contiguous()
+        shift_labels = labels[..., 1:].contiguous()
+
+        shift_logits = shift_logits.view(-1, shift_logits.size(-1))
+        shift_labels = shift_labels.view(-1)
+
+        loss = self.loss_fn(shift_logits, shift_labels)
+
+        true_lengths = torch.sum(labels != 0, dim=-1).float()
+        pred_lengths = torch.sum(torch.argmax(logits, dim=-1) != 0, dim=-1).float()
```

Comment:
why exactly are we checking for `0` here?

Reply:
The idea was to ignore the PAD token Ids which we generally apply on answers text (max_length_answer). In models like Llama2, etc. Pad token id is 0, so I took it as default. But yeah, it could be possible that token ID may differ for different tokenizers.

Reply:
Yeah this will not be the case all the time. Actually, most of the time we set it to eos token. This needs to be dynamically set. Actually, it might be better to use the `attention_mask` for it. Similar logic like in batch padding where we are also calculating the length: https://github.com/h2oai/h2o-llmstudio/blob/8d1c13684bc5e173b635e4d7867fa84d31c5998e/llm_studio/src/utils/data_utils.py#L640

Reply:
Yes right, Thanks for pointing out. I will update the loss function based on the tokenizer.

Reply:
Most should be in the `cfg` already stored, can look at the `get_tokenizer` function.